File: py/db/pymongo-example/utils/mongo.py
# pymongo
from pymongo import ReturnDocument
# connection
from connections.mongo import db
# standard library
from contextlib import contextmanager
# package
import arrow

# collections

ninjas = db.ninjas
jutsus = db.jutsus

# DAL class


class MongoDAL():

    @staticmethod
    def get_document_from_result(result):
        document = result.ops[0]
        return document

    # ---
    # ninjas
    # ---

    def get_ninja(self, _id):
        ninja = ninjas.find_one({'_id': _id})
        return ninja

    def get_ninjas(self, conditions):
        ninjas = ninjas.find(conditions)
        return ninjas

    def insert_ninja(self, ninja):
        now = arrow.utcnow().format()
        result = ninjas.insert_one({
            **ninja,
            'createdAt': now
        })
        inserted_ninja = ninjas.find_one({'_id': result.inserted_id})
        return inserted_ninja

    def update_ninja(self, _id, updates):
        now = arrow.utcnow().format()
        updated_ninja = ninjas.find_one_and_update(
            filter={'_id': _id},
            update={'$set': {**updates, 'updatedAt': now}},
            return_document=ReturnDocument.AFTER
        )
        return updated_ninja

    def delete_ninja(self, _id):
        deleted_ninja = find_one_and_delete({'_id': _id})
        return deleted_ninja

    # ---
    # jutsus
    # ---

    def get_jutsu(self, _id):
        jutsu = jutsus.find_one({'_id': _id})
        return jutsu

    def get_jutsus(self, conditions):
        jutsus = jutsus.find(conditions)
        return jutsus

    def insert_jutsu(self, jutsu):
        now = arrow.utcnow().format()
        result = jutsus.insert_one({
            **jutsu,
            'createdAt': now
        })
        inserted_jutsu = jutsus.find_one({'_id': result.inserted_id})
        return inserted_jutsu

    def update_jutsu(self, _id, updates):
        now = arrow.utcnow().format()
        updated_jutsu = jutsus.find_one_and_update(
            filter={'_id': _id},
            update={'$set': {**updates, 'updatedAt': now}},
            return_document=ReturnDocument.AFTER
        )
        return updated_jutsu

    def delete_jutsu(self, _id):
        deleted_jutsu = find_one_and_delete({'_id': _id})
        return deleted_jutsu

    # ---
    # ninjas_jutsus
    # ---

    # query many-to-many
    # https://stackoverflow.com/a/39476690

    def get_ninja_with_related_jutsu(self, _id):
        ninja_with_related_jutsu = ninjas.aggregate([
            # condition
            {"$match": {'_id': _id}},
            # with related
            {'$lookup': {
                'from': 'jutsus',
                'localField': 'jutsuIds',
                'foreignField': '_id',
                'as': 'jutsus'
            }}
        ])
        return list(ninja_with_related_jutsu)

    def get_jutsu_with_related_ninja(self, _id):
        jutsu_with_related_ninja = jutsus.aggregate([
            # condition
            {"$match": {'_id': _id}},
            # with related
            {'$lookup': {
                'from': 'ninjas',
                'localField': 'ninjaIds',
                'foreignField': '_id',
                'as': 'ninjas'
            }}
        ])
        return list(jutsu_with_related_ninja)

    def add_known_jutsu(self, ninja_id, jutsu_id):
        now = arrow.utcnow().format()
        ninjas.find_one_and_update(
            {'_id': ninja_id},
            {'$push': {'jutsuIds': jutsu_id}, '$set': {'updatedAt': now}},
        )
        jutsus.find_one_and_update(
            {'_id': jutsu_id},
            {'$push': {'ninjaIds': ninja_id}, '$set': {'updatedAt': now}},
        )

    def remove_known_jutsu(self, ninja_id, jutsu_id):
        now = arrow.utcnow().format()
        ninjas.find_one_and_update(
            {'_id': ninja_id},
            {'$pull': {'jutsuIds': jutsu_id}, '$set': {'updatedAt': now}},
        )
        jutsus.find_one_and_update(
            {'_id': jutsu_id},
            {'$pull': {'ninjaIds': ninja_id}, '$set': {'updatedAt': now}},
        )

File: js/ts/db/mongodb-notes/utils/mongo.ts
import mongodb, { Db } from "mongodb";
import { getConnection } from "../connections/mongo";
import {
  Ninja,
  NinjaConditions,
  NinjaNew,
  NinjaUpdates,
  Jutsu,
  JutsuConditions,
  JutsuNew,
  JutsuUpdates,
} from "../types";

// ---
// collections
// ---

const collections = {
  NINJAS: "ninjas",
  JUTSUS: "jutsus",
};

// ---
// ninjas
// ---

const getNinja = async (_id: mongodb.ObjectID): Promise<Ninja> => {
  const db = await getConnection();
  const ninja = await db.collection(collections.NINJAS).findOne({ _id });
  return ninja;
};

const getNinjas = async (conditions: NinjaConditions): Promise<Ninja[]> => {
  const db = await getConnection();
  const ninjas = await db.collection(collections.NINJAS).find(conditions).toArray();
  return ninjas;
};

const insertNinja = async (ninja: NinjaNew): Promise<Ninja> => {
  const db = await getConnection();
  const result = await db.collection(collections.NINJAS).insertOne({
    ...ninja,
    createdAt: new Date(),
  });
  const insertedNinja: Ninja = await db
    .collection(collections.NINJAS)
    .findOne({ _id: result.insertedId });
  return insertedNinja;
};

const updateNinja = async (_id: mongodb.ObjectID, updates: NinjaUpdates): Promise<Ninja> => {
  const db = await getConnection();
  const now = new Date();
  const updatedNinja = (await db.collection(collections.NINJAS).findOneAndUpdate(
    { _id }, // filter
    { $set: { ...updates, updatedAt: now } }, // updates
    { returnOriginal: false }, // options
  )) as Ninja;
  return updatedNinja;
};

const deleteNinja = async (_id: mongodb.ObjectID): Promise<Ninja> => {
  const db = await getConnection();
  const deletedNinja = (await db.collection(collections.NINJAS).findOneAndDelete({ _id })) as Ninja;
  return deletedNinja;
};

// ---
// jutsus
// ---

const getJutsu = async (_id: mongodb.ObjectID): Promise<Jutsu> => {
  const db = await getConnection();
  const jutsu = await db.collection(collections.JUTSUS).findOne({ _id });
  return jutsu;
};

const getJutsus = async (conditions: JutsuConditions): Promise<Jutsu[]> => {
  const db = await getConnection();
  const jutsus = await db.collection(collections.JUTSUS).find(conditions).toArray();
  return jutsus;
};

const insertJutsu = async (jutsu: JutsuNew): Promise<Jutsu> => {
  const db = await getConnection();
  const result = await db.collection(collections.JUTSUS).insertOne({
    ...jutsu,
    createdAt: new Date(),
  });
  const insertedJutsu: Jutsu = await db
    .collection(collections.JUTSUS)
    .findOne({ _id: result.insertedId });
  return insertedJutsu;
};

const updateJutsu = async (_id: mongodb.ObjectID, updates: JutsuUpdates): Promise<Jutsu> => {
  const db = await getConnection();
  const now = new Date();
  const response = (await db.collection(collections.JUTSUS).findOneAndUpdate(
    { _id }, // filter
    { $set: { ...updates, updatedAt: now } }, // updates
    { returnOriginal: false }, // options
  )) as Jutsu;
  return response;
};

const deleteJutsu = async (_id: mongodb.ObjectID): Promise<Jutsu> => {
  const db = await getConnection();
  const deletedJutsu = (await db.collection(collections.JUTSUS).findOneAndDelete({ _id })) as Jutsu;
  return deletedJutsu;
};

// ---
// ninjas_jutsus
// ---

const getNinjaWithRelatedJutsus = async (_id: mongodb.ObjectID): Promise<Ninja> => {
  const db = await getConnection();
  const [ninjaWithRelatedJutsus] = await db
    .collection(collections.NINJAS)
    .aggregate([
      { $match: { _id } }, // condition
      {
        $lookup: {
          from: "jutsus",
          localField: "jutsuIds",
          foreignField: "_id",
          as: "jutsus",
        },
      },
    ])
    .toArray();
  return ninjaWithRelatedJutsus;
};

const getJutsuWithRelatedNinjas = async (_id: mongodb.ObjectID): Promise<Jutsu> => {
  const db = await getConnection();
  const [jutsuWithRelatedNinjas] = await db
    .collection(collections.NINJAS)
    .aggregate([
      { $match: { _id } }, // condition
      {
        $lookup: {
          from: "ninjas",
          localField: "ninjaIds",
          foreignField: "_id",
          as: "ninjas",
        },
      },
    ])
    .toArray();
  return jutsuWithRelatedNinjas;
};

const addKnownJutsu = async (
  ninjaId: mongodb.ObjectID,
  jutsuId: mongodb.ObjectID,
): Promise<void> => {
  const db = await getConnection();
  const now = new Date();
  await db.collection(collections.NINJAS).findOneAndUpdate(
    { _id: ninjaId }, // filter
    { $push: { jutsuIds: jutsuId }, $set: { updatedAt: now } }, // updates
  );
  await db.collection(collections.JUTSUS).findOneAndUpdate(
    { _id: jutsuId }, // filter
    { $push: { ninjaIds: ninjaId }, $set: { updatedAt: now } }, // updates
  );
};

const removeKnownJutsu = async (
  ninjaId: mongodb.ObjectID,
  jutsuId: mongodb.ObjectID,
): Promise<void> => {
  const db = await getConnection();
  const now = new Date();
  await db
    .collection(collections.NINJAS)
    .findOneAndUpdate({ _id: ninjaId }, { $pull: { jutsuIds: jutsuId }, $set: { updatedAt: now } });
  await db
    .collection(collections.JUTSUS)
    .findOneAndUpdate({ _id: jutsuId }, { $pull: { ninjaIds: ninjaId }, $set: { updatedAt: now } });
};

// ---
// export
// ---

const mongoDAL = {
  // ninjas
  getNinja,
  getNinjas,
  insertNinja,
  updateNinja,
  deleteNinja,
  // jutsus
  getJutsu,
  getJutsus,
  insertJutsu,
  updateJutsu,
  deleteJutsu,
  // ninjas_jutsus
  addKnownJutsu,
  removeKnownJutsu,
  getNinjaWithRelatedJutsus,
  getJutsuWithRelatedNinjas,
};

export default mongoDAL;
